libs/crypto/merkle-impl/src/main/kotlin/net/corda/crypto/merkle/impl/MerkleProofImpl.kt

```diff
@@ -226,7 +223,52 @@ class MerkleProofImpl(
                 "MerkleProof root hash calculation ended with ${nodeHashes.size} node hashes instead of one."
             )
         }
-        return nodeHashes.single().second
+        return nodeHashes.single().hash
+    }
+
+    /**
+     *
+     * Work out a Merkle proof with specified leaves.
+     *
+     * Throws IllegalArgumentException if some requested leaf indices are not available in the source proof,
+     * or if no leaf indices are requested.
+     *
+     * @param leafIndices indices  of the known leaves to include in the output proof
+     * @return A new Merkle proof covering the specified leaves.
+     */
+    fun subset(digest: MerkleTreeHashDigest, leafIndices: List<Int>): MerkleProofImpl {
+        val outLeaves = leaves.filter { it.index in leafIndices }
+        require(outLeaves.size == leafIndices.size) { "some leaves are not available in input proof"}
+        require(outLeaves.isNotEmpty()) { "output proof must have at least one known leaf"}
+        // we will build up a set of available leaves as we track. This will start out with the known leaves
+        // in the output subset proof, and will be augmented as we decide to add output hashes to the proof
+        val availableLeaves = leafIndices.toMutableSet()
+        // We work out the hashes for the new subset proof by considering, for the original proof, each hash that
+        // is calculated when we verify the proof by calculating the root.
+        val outHashes: MutableList<SecureHash> = mutableListOf()
+        val treeDepth = MerkleTreeImpl.treeDepth(treeSize)
+        calculateRootInstrumented(digest) { info ->
+            val height = treeDepth - info.level // how many levels above the leaves, 0 for being at the leaf
+            val leftmostLeaf = info.node.indexWithinLevel shl height
+            val afterLeaf = min(leftmostLeaf + (1 shl height), treeSize)
+            val unknowns = (leftmostLeaf until afterLeaf).any { it !in availableLeaves }
+            val allUnknown = (leftmostLeaf until afterLeaf).all { it !in availableLeaves }
+            // work out the sibling node i.e. the node which shares a direct parent node with this one
+            val siblingIndex = info.node.indexWithinLevel xor 1
+            val siblingLeftmostLeaf = siblingIndex shl height
+            val afterSiblingLeaf = min((siblingIndex + 1) shl height, treeSize)
```

Comment:
This calculation is not fully symmetric with https://github.com/corda/corda-runtime-os/pull/5393/files#diff-097e799c831ce16be1cfa92a78a730a5eb62fb2e03183e3e093a714c817ab358R252-R253
Is this intentional?
```suggestion
            val siblingLeftmostLeaf = siblingIndex shl height
            val afterSiblingLeaf = min(siblingLeftmostLeaf + (1 shl height), treeSize)
```
Both look correct for me and doing the same, just maybe it would be nicer to do that in the same way.

Reply:
I think that's a bit more confusing.

Reply:
Then, this approach in the other place too?

Reply:
Oh you mean
>             val afterLeaf = min(leftmostLeaf + (1 shl height), treeSize)

and
>             val afterSiblingLeaf = min((siblingIndex + 1) shl height, treeSize)

they don't actually have the same value, I did check, what's here is correct. But maybe could use more comments.

Reply:
So for instance
```
>>> print( (5+1) << 3)
48
>>> print ( 5 << (1+3))
80
```

i.e. left bitshift does not have the associativity you think it does.

Reply:
I meant that `afterSiblingLeaf` and `afterLeaf` are derived differently from their inputs.
Their expression structures and operators could be the same. (with different arguments)